File: monitor/monitor.js
'use strict';

// html page widget
var loginBtn = document.querySelector('#loginBtn');
var hangUpBtn = document.querySelector('#hangUpBtn');
var localVideo = document.querySelector('#localVideo');
hangUpBtn.disabled = true;

// signalling server IP address
var serverURL = 'ws://47.115.205.65:9091'
// monitor username in server, can be used as client password
var myUsername = 'monitor';
// client username in server
var connectedUsername = null;
// WebSocket object
var myWebSocket;
// RTCPeerConnection object
var myPeerConnection;
// camera stream
var stream;
// whether RTCPeerConnection has been created
var RTCPeerConnectionCreated = false;

// Set stream Codec
const codecPreferences = document.querySelector('#codecPreferences');
const supportsSetCodecPreferences = window.RTCRtpTransceiver &&
    'setCodecPreferences' in window.RTCRtpTransceiver.prototype;

// setting the STUN/TURN server IP
const configuration = {
    "iceServers":
        [
            {
                'urls': 'stun:47.115.205.65:3478',
            },
            {
                'urls': 'turn:47.115.205.65',
                'username': 'hhr',
                'credential': '123456'
            }
        ]
};

/**
 * botton to start WebSocket connection and open the camera
 */
loginBtn.addEventListener("click", function () {
    // connect to signalling server with WebSocket
    myWebSocket = new WebSocket(serverURL);
    // message handle function
    myWebSocket.onmessage = function (message) {
        // print message in console
        console.log("Got message: ", message.data);
        // handle message
        var data = JSON.parse(message.data);
        console.log(data)
        switch (data.type) {
            case "login":
                handleLogin(data.success);
                break;
            // call from client
            case "offer":
                handleOffer(data.offer, data.name);
                break;
            case "answer":
                handleAnswer(data.answer);
                break;
            // receive ICE candidate
            case "candidate":
                handleCandidate(data.candidate);
                break;
            // client disconnect
            case "leave":
                handleLeave();
                break;
            // client require monitor video
            case "cmd":
                handleCmd(data.sender);
                break;
            // server information
            case "sendinfo":
                alert(data.info);
                if (data.close == true) {
                    window.location.href = "about:blank";
                }
                break;
            default:
                break;
        }
    };
    // error handle function
    myWebSocket.onerror = function (err) {
        console.log("Got error: ", err);
    };
    // connection handle function
    myWebSocket.onopen = function () {
        console.log("Connected to the signalling server.");
        // login to the signalling server
        if (myUsername.length > 0) {
            sendToServer({
                type: "login",
                name: myUsername
            });
        }
        //getting local video stream
        navigator.mediaDevices.getUserMedia({
            video: true, audio: true
        }).then(
            myStream => {
                stream = myStream
                //displaying local video stream on the page
                localVideo.srcObject = stream;
                window.localStream = stream;
            }
        ).catch(
            err => {
                console.log(err);
            }
        );
        if (supportsSetCodecPreferences) {
            const { codecs } = RTCRtpSender.getCapabilities('video');
            console.log('RTCRtpSender.getCapabilities(video):\n', codecs);
            codecs.forEach(codec => {
                if (['video/red', 'video/ulpfec', 'video/rtx'].includes(codec.mimeType)) {
                    return;
                }
                const option = document.createElement('option');
                option.value = (codec.mimeType + ' ' + (codec.sdpFmtpLine || '')).trim();
                option.innerText = option.value;
                codecPreferences.appendChild(option);
            });
            codecPreferences.disabled = false;
        } else {
            console.warn('Your browser do not support changing codec.');
        }
    };
    // enable/disable corresponding button
    loginBtn.disabled = true;
    hangUpBtn.disabled = false;
});

/**
 * botton to disconnect to the client/server
 */
hangUpBtn.addEventListener("click", function () {
    // inform the server we are going to hangup
    sendToServer({
        type: "sendinfo",
        info: "monitor disconnected to server",
        close: false
    });
    // inform client we have hangup
    if (connectedUsername == null) {
        sendToServer({
            type: "leave",
            name: 'undefined'
        });
    } else {
        sendToServer({
            type: "leave",
        });
    }
    // if client has connected, close the RTCPeerConnection, otherwise, just stop getting stream from camera
    if (connectedUsername != null) {
        handleLeave();
    }
    // stop getting local video stream
    stream.getTracks().forEach(track => track.stop());
    localVideo.srcObject = null;
    window.localStream = null;
    // enable/disable corresponding button
    hangUpBtn.disabled = true;
    loginBtn.disabled = false;
});

/**
 * send message to server
 * @param message
 */
function sendToServer(message) {
    // if client connected, send message to client through server
    if (connectedUsername) {
        message.name = connectedUsername;
    }
    myWebSocket.send(JSON.stringify(message));
};

/**
 * initialize the RTCPeerConnection
 * @returns 
 */
function initPeer() {
    // add prefix
    let PeerConnection = window.RTCPeerConnection ||
        window.mozRTCPeerConnection ||
        window.webkitRTCPeerConnection;

    try {
        myPeerConnection = new RTCPeerConnection(configuration);
        // add stream to track
        if ("addTrack" in myPeerConnection) {
            stream.getTracks().forEach(track => {
                myPeerConnection.addTrack(track, stream);
            });
        } else {
            myPeerConnection.addStream(stream);
        }

        if (supportsSetCodecPreferences) {
            // get available codec
            const preferredCodec = codecPreferences.options[codecPreferences.selectedIndex];
            if (preferredCodec.value !== '') {
                const [mimeType, sdpFmtpLine] = preferredCodec.value.split(' ');
                const { codecs } = RTCRtpSender.getCapabilities('video');
                const selectedCodecIndex = codecs.findIndex(c => c.mimeType === mimeType && c.sdpFmtpLine === sdpFmtpLine);
                const selectedCodec = codecs[selectedCodecIndex];
                codecs.splice(selectedCodecIndex, 1);
                codecs.unshift(selectedCodec);
                console.log(codecs);
                const transceiver = myPeerConnection.getTransceivers().find(t => t.sender && t.sender.track === localStream.getVideoTracks()[0]);
                transceiver.setCodecPreferences(codecs);
                console.log('Preferred Codec: ', selectedCodec);
            }
        }
        codecPreferences.disabled = true;

        // setup stream listening
        if ("ontrack" in myPeerConnection) {
            // when a remote user adds stream to the peer connection, we display it
            myPeerConnection.ontrack = async function (event) {
                hangUpBtn.disabled = false;
            }

        } else {
            // when a remote user adds stream to the peer connection, we display it
            myPeerConnection.onaddstream = async function (event) {
                hangUpBtn.disabled = false;
            }

        }

        myPeerConnection.oniceconnectionstatechange = async function (event) {
            // handle changes of ICE connection state
            console.log("ICE connection state changed to " + myPeerConnection.iceConnectionState);
            switch (myPeerConnection.iceConnectionState) {
                case "closed":
                case "failed":
                case "disconnected":
                    handleLeave();
                    break;
                case "connected":
                    hangUpBtn.disabled = false;
                    break;
            }
        };
        myPeerConnection.onicegatheringstatechange = async function (event) {
            console.log("ICE gathering state changed to " + myPeerConnection.iceGatheringState);
        }
        myPeerConnection.onsignalingstatechange = async function (event) {
            // setup a |signalingstatechange| event handler. This will detect when the signaling connection is closed.
            if (myPeerConnection == null) {
                return;
            }
            console.log("WebRTC signalling state changed to " + myPeerConnection.signalingState);
            switch (myPeerConnection.signalingState) {
                case "closed":
                    handleLeave();
                    break;
            }
        }
        // RTCPeerConnectionCreated has created
        RTCPeerConnectionCreated = true;
    } catch (err) {
        console.log('Failed to create PeerConnection: ' + err.message);
        alert('Cannot create RTCPeerConnection object.');
        RTCPeerConnectionCreated = false;
        return;
    }

    // display codec actually used
    setTimeout(async () => {
        const stats = await myPeerConnection.getStats();
        stats.forEach(stat => {
            if (!(stat.type === 'outbound-rtp' && stat.kind === 'video')) {
                return;
            }
            const codec = stats.get(stat.codecId);
            console.log('Actually using ' + codec.mimeType +
                ' ' + (codec.sdpFmtpLine ? codec.sdpFmtpLine + ' ' : '') +
                ', payloadType=' + codec.payloadType + '. Encoder: ' + stat.encoderImplementation);
        });
    }, 1000);
}

/**
 * when server sends login reply
 * @param success
 */
function handleLogin(success) {
    // client username is a duplicate of myUsername
    if (success === false) {
        alert("The same username has in server, try another username.");
    }
};

/**
 * when client sends us an offer
 * @param offer offer content
 * @param name  username of whom send offer
 * @returns {Promise<void>}
 */
async function handleOffer(offer, name) {
    // setup a new connection if RTCPeerConnection hasn't been created
    if (RTCPeerConnectionCreated == false) {
        initPeer();
    }

    connectedUsername = name;
    // handle offer
    await myPeerConnection.setRemoteDescription(new RTCSessionDescription(offer));
    // create answer for the offer
    myPeerConnection.createAnswer().then(function (answer) {

        myPeerConnection.setLocalDescription(answer);
        sendToServer({
            type: "answer",
            answer: answer
        });
    }).catch(function (error) {
        alert("Error when creating an answer.");
    });
};

/**
 * when we get an answer from client
 * @param answer answer from client
 * @returns {Promise<void>}
 */
async function handleAnswer(answer) {
    // handle answer
    await myPeerConnection.setRemoteDescription(new RTCSessionDescription(answer));
};

/**
 * when we get an ICE candidate from STUN/TURN server
 * @param candidate ICE candidate
 */
function handleCandidate(candidate) {
    myPeerConnection.addIceCandidate(new RTCIceCandidate(candidate));
};

/**
 * when monitor disconnect
 */
function handleLeave() {

    // clear client username
    connectedUsername = null;
    // disconnect all event listeners
    myPeerConnection.onicecandidate = null;
    myPeerConnection.onaddstream = null;
    myPeerConnection.ontrack = null;
    myPeerConnection.onsignalingstatechange = null;
    myPeerConnection.onicegatheringstatechange = null;
    myPeerConnection.close();
    myPeerConnection = null;
    // RTCPeerConnection not created
    RTCPeerConnectionCreated = false;
    // disable hangup botton on the page
    hangUpBtn.disabled = false;

};

/**
 * when client requires our video stream
 * @param {*} sender username of client
 * @returns 
 */
function handleCmd(sender) {
    // create Offer and setLocalDescription
    if (sender.length > 0) {
        // username of our stream's receiver
        connectedUsername = sender;
        // inform client that monitor is not opened
        if (stream == null || !stream.active) {
            sendToServer({
                type: "sendinfo",
                info: "The monitor has not been opened.",
                close: false
            });
            return;
        }
        // initialize RTCPeerConnection
        initPeer();
        // create an offer
        myPeerConnection.createOffer().then(function (offer) {
            myPeerConnection.setLocalDescription(offer);
            sendToServer({
                type: "offer",
                offer: offer
            });
        }).catch(function (error) {
            alert("Error when creating an offer.");
        });
    }
}

File: monitor/monitor.py
import websockets
import aiortc
from aiortc.contrib.media import MediaPlayer
from aiortc.sdp import candidate_from_sdp
import json
import asyncio
import time

# signaling server IP address
serverURL = "ws://47.115.205.65:9091"
# monitor ID in server
myID = "monitor"
# client ID in server
connectedID = None
# RTCPeerConnection object
myPeerConnection = None
# audio and video track
audioTrack = None
videoTrack = None
# whether RTCPeerConnection has been created
RTCPeerConnectionCreated = False


async def sendToServer(myWebSocket, message: dict):
    # if client connected, send message to client through server
    if connectedID:
        # add receiver's ID to the message
        message.update({"ID": connectedID})
    # change dict to JSON string
    message = json.dumps(message)
    # send message
    await myWebSocket.send(message)


def initPC():
    global myPeerConnection
    global RTCPeerConnectionCreated
    # configure STUN/TURN server
    STUNServer = aiortc.RTCIceServer(urls="stun:47.115.205.65:3478")
    TURNServer = aiortc.RTCIceServer(urls="urls",
                                     username="username",
                                     credential="credential")
    configuration = aiortc.RTCConfiguration(
        iceServers=[STUNServer, TURNServer])
    # initialize RTCPeerConnection
    try:
        # create RTCPeerConnection and add track to be sent
        myPeerConnection = aiortc.RTCPeerConnection(
            configuration=configuration)
        if audioTrack != None:
            myPeerConnection.addTrack(audioTrack)
        if videoTrack != None:
            setCodec(myPeerConnection.addTrack(videoTrack), "video/H264")

        # add event listener
        @myPeerConnection.on("connectionstatechange")
        async def on_connectionstatechange():
            print("Connection state is %s." % myPeerConnection.connectionState)

        @myPeerConnection.on("iceconnectionstatechange")
        async def on_iceconnectionstatechange():
            print("ICE connection state is %s." %
                  myPeerConnection.iceConnectionState)
            if myPeerConnection.iceConnectionState == "closed":
                await handleLeave()

        @myPeerConnection.on("icegatheringstatechange")
        async def on_icegatheringstatechange():
            print("ICE gathering state is %s." %
                  myPeerConnection.iceGatheringState)

        @myPeerConnection.on("signalingstatechange")
        async def on_signalingstatechange():
            if myPeerConnection == None:
                return
            print("Signaling state is %s." % myPeerConnection.signalingState)
            if myPeerConnection.signalingState == "closed":
                await handleLeave()

        # create RTCPeerConnection successfully
        RTCPeerConnectionCreated = True
    except Exception as e:
        # create RTCPeerConnection failed
        print("Error occurred when creating RTCPeerConnection.")
        print(e, e.__traceback__.tb_lineno)


def setCodec(sender: aiortc.rtcpeerconnection.RTCRtpSender, forcedCodec: str):
    global myPeerConnection
    # assign codec used in RTCPeerConnection
    kind = forcedCodec.split("/")[0]
    codecs = aiortc.RTCRtpSender.getCapabilities(kind).codecs
    transceiver = next(t for t in myPeerConnection.getTransceivers()
                       if t.sender == sender)
    transceiver.setCodecPreferences(
        [codec for codec in codecs if codec.mimeType == forcedCodec])


def handleLogin(message: dict):
    # handle connection error
    if message["success"] == False:
        print("ID of WebCam has been used, try another ID.")


async def handleAnswer(message: dict):
    global myPeerConnection
    # handle answer string
    sdp = message["answer"]["sdp"]
    typ = message["answer"]["type"]
    try:
        await myPeerConnection.setRemoteDescription(
            aiortc.RTCSessionDescription(sdp=sdp, type=typ))
    except Exception as e:
        print("Error occurred when setting Remote Description.")
        print(e, e.__traceback__.tb_lineno)


async def handleCandidate(message: dict):
    global myPeerConnection
    # handle candidate string
    try:
        # remove candidate prefix
        candidate = message["candidate"]["candidate"][10:]
        # split candidate string by space
        candidate = candidate.split()
        # assign parameters
        component = int(candidate[1])
        foundation = candidate[0]
        ip = candidate[4]
        port = int(candidate[5])
        priority = int(candidate[3])
        protocol = candidate[2]
        typ = candidate[7]
        if "raddr" in candidate:
            relatedAddress = candidate[9]
        else:
            relatedAddress = None
        if "rport" in candidate:
            relatedPort = candidate[11]
        else:
            relatedPort = None
        sdpMid = message["candidate"]["sdpMid"]
        sdpMLineIndex = message["candidate"]["sdpMLineIndex"]
    except Exception as e:
        print("Error occurred when handling candidate string.")
        print(e, e.__traceback__.tb_lineno)
    # add ICE candidate
    try:
        await myPeerConnection.addIceCandidate(
            aiortc.RTCIceCandidate(component=component,
                                   foundation=foundation,
                                   ip=ip,
                                   port=port,
                                   priority=priority,
                                   protocol=protocol,
                                   type=typ,
                                   relatedAddress=relatedAddress,
                                   relatedPort=relatedPort,
                                   sdpMid=sdpMid,
                                   sdpMLineIndex=sdpMLineIndex))
    except Exception as e:
        print("Error occurred when adding ICE candidate.")
        print(e, e.__traceback__.tb_lineno)


async def handleCall(myWebSocket, message: dict):
    global connectedID
    global myPeerConnection
    global audioTrack
    global videoTrack
    # create Offer and setLocalDescription
    if len(message["receiver"]) > 0:
        # username of our stream's receiver
        connectedID = message["receiver"]
        # when client called, try to get video/audio stream
        try:
            # get camera
            videoTrack = MediaPlayer("/dev/video0",
                                     format="v4l2",
                                     options={
                                         "video_size": "640x480"
                                     }).video
            # get microphone
            audioTrack = MediaPlayer(file="hw:0,0",
                                     format="alsa",
                                     options={
                                         "channels": "1"
                                     }).audio
        except Exception as e:
            print("Error occurred when opening camera or microphone.")
            print(e, e.__traceback__.tb_lineno)
        # inform client that monitor is not opened
        if audioTrack and videoTrack == None:
            await sendToServer(
                myWebSocket, {
                    "type": "sendinfo",
                    "info": "The monitor has not been opened.",
                    "close": False
                })
            return
        # initialize RTCPeerConnection
        initPC()
        # create an offer
        try:
            # create offer to set up RTCPeerConnection
            offer = await myPeerConnection.createOffer()
            await myPeerConnection.setLocalDescription(offer)
            """
            It's strange that "sdp" in "offer" doesn't include self IP address, 
            so here use "sdp" in RTCPeerConnection.localDescription instead.
            """
            await sendToServer(
                myWebSocket, {
                    "type": "offer",
                    "offer": {
                        "type": offer.type,
                        "sdp": myPeerConnection.localDescription.sdp
                    }
                })
        except Exception as e:
            print("Error occurred when creating offer.")
            print(e, e.__traceback__.tb_lineno)


async def handleLeave():
    global myPeerConnection
    global RTCPeerConnectionCreated
    global connectedID
    # clear client username
    connectedID = None
    # disconnect RTCPeerConnection
    await myPeerConnection.close()
    RTCPeerConnectionCreated = False


async def onOpen(myWebSocket):
    # callback function when WebSocket opened
    # receive the information of successfully start WebSocket
    success = await myWebSocket.recv()
    print(success)
    # send connect information to the server
    await sendToServer(myWebSocket, {
        "type": "login",
        "ID": myID,
        "username": "0000",
        "password": "0000"
    })


async def onMessage(myWebSocket, message: str):
    # callback function when got message from server
    print("Got message: %s" % message)
    # load JSON string
    message = json.loads(message)
    # choose handle function according to the type of message
    messageType = message["type"]
    if (messageType == "login"):
        handleLogin(message)
    # receive answer from client
    elif (messageType == "answer"):
        await handleAnswer(message)
    # receive ICE candidate from client
    elif (messageType == "candidate"):
        await handleCandidate(message)
    # called by client, raise RTCPeerConnection
    elif (messageType == "call"):
        await handleCall(myWebSocket, message)
    # client disconnected
    elif (messageType == "leave"):
        await handleLeave()


async def onClose(myWebSocket):
    global audioTrack
    global videoTrack
    # inform website client we are going to hang up
    await sendToServer(
        myWebSocket, {
            "type": "sendinfo",
            "info": "Monitor disconnected to server.",
            "close": False
        })
    # log out from the server
    if connectedID == None:
        await sendToServer(myWebSocket, {"type": "leave", "ID": "undefined"})
    else:
        await sendToServer(myWebSocket, {"type": "leave"})
        # if has connected to client, close the RTCPeerConnection, otherwise, just stop getting stream
        await handleLeave()
    audioTrack = None
    videoTrack = None


async def WS():
    # connect to server through WebSocket
    print("Connecting to server...")
    while True:
        try:
            async with websockets.connect(serverURL) as myWebSocket:
                # callback function when connect to server successfully
                await onOpen(myWebSocket)
                while True:
                    try:
                        # receive message from server
                        message = await myWebSocket.recv()
                        # callback function to handle message
                        await onMessage(myWebSocket, message)
                    except KeyboardInterrupt as e:
                        # close program when keyboard interrupt
                        await onClose(myWebSocket)
                        print("Program closed due to keyboard interrupt.")
                        print(e, e.__traceback__.tb_lineno)
                        return
                    except websockets.exceptions.ConnectionClosedError as e:
                        print(
                            "Connection closed due to error occurred. Reconnection will start after 5s."
                        )
                        print(e, e.__traceback__.tb_lineno)
                        time.sleep(5)
                        break
        except Exception as e:
            print(
                "WebSocket cannot connect to server. Reconnection will start after 5s."
            )
            print(e, e.__traceback__.tb_lineno)
            time.sleep(5)


if __name__ == "__main__":
    asyncio.get_event_loop().run_until_complete(WS())
